tests/test_cases/test_axi/test_axi.py

```diff
@@ -0,0 +1,349 @@
+        ('araddr', 0x00), ('rdata', 0x00),
+        ('araddr', 0x10), ('rdata', 0x00),
+        ('araddr', 0x20), ('rdata', 0x00), ('rdata', 0x00), ('rdata', 0x00),
+    ])
+
+    AXI4MemoryMap(1024, dut, "MAXI", dut.Clk, arready_delay=-1,
+                  rresp_delay=10, max_threads=1)
+    bus = get_bus(dut)
+
+    yield reset(dut)
+
+    a = yield read(dut.Clk, bus, 0x00, 1, thread_id=0x00)
+    yield a.join()
+    b = yield read(dut.Clk, bus, 0x10, 1, thread_id=0x00)
+    yield b.join()
+    c = yield read(dut.Clk, bus, 0x20, 3, thread_id=0x00)
+    yield c.join()
+
+    yield ClockCycles(dut.Clk, 3)
+
+
+@cocotb.test()
+def test_read_multithread(dut):
+    """Test Master AXI Read
+    """
+    mon = AXI4(dut, "MAXI", dut.Clk, reset=dut.Reset)
+    sb = Scoreboard(dut, fail_immediately=False)
+    sb.add_interface(mon, [
+        ('araddr', 0x00), ('rdata', 0x00), ('rdata', 0x00), ('rdata', 0x00),
+        ('araddr', 0x10), ('rdata', 0x00), ('rdata', 0x00), ('rdata', 0x00),
+        ('araddr', 0x20), ('rdata', 0x00), ('rdata', 0x00), ('rdata', 0x00),
+        ('araddr', 0x30), ('rdata', 0x00), ('rdata', 0x00), ('rdata', 0x00),
+    ], reorder_depth=5)
+
+    AXI4MemoryMap(1024, dut, "MAXI", dut.Clk, arready_delay=-1,
+                  rresp_delay=10, max_threads=2)
+    bus = get_bus(dut)
+
+    yield reset(dut)
+
+    a = yield read(dut.Clk, bus, 0x00, 3, thread_id=0x00)
+    b = yield read(dut.Clk, bus, 0x10, 3, thread_id=0x01)
+    c = yield read(dut.Clk, bus, 0x20, 3, thread_id=0x02)
+    d = yield read(dut.Clk, bus, 0x30, 3, thread_id=0x03)
+
+    yield a.join()
+    yield b.join()
+    yield c.join()
+    yield d.join()
+
+    yield ClockCycles(dut.Clk, 3)
+
+
+@cocotb.test()
+def test_read_write(dut):
+    """Test Master AXI Read and Writing
+    """
+    mon = AXI4(dut, "MAXI", dut.Clk, reset=dut.Reset)
+    sb = Scoreboard(dut, fail_immediately=False)
+    sb.add_interface(mon, [
+        ('awaddr', 0x00), ('wdata', 0xAA), ('bid', 0x00),
+        ('araddr', 0x00), ('rdata', 0xAA), ('rdata', 0x00), ('rdata', 0x00),
+
+        ('awaddr', 0x08), ('wdata', 0xBB), ('bid', 0x01),
+        ('araddr', 0x00), ('rdata', 0xAA), ('rdata', 0xBB), ('rdata', 0x00),
+
+        ('awaddr', 0x00), ('wdata', 0xCC), ('wdata', 0xCC), ('wdata', 0xCC),
+        ('bid', 0x02),
+        ('araddr', 0x00), ('rdata', 0xCC), ('rdata', 0xCC), ('rdata', 0xCC),
+    ], reorder_depth=5)
+
+    AXI4MemoryMap(1024, dut, "MAXI", dut.Clk, awready_delay=-1,
+                  wready_delay=-1, bresp_delay=10, arready_delay=-1,
+                  rresp_delay=3, max_threads=1)
+    bus = get_bus(dut)
+
+    yield reset(dut)
+
+    a = yield write(dut.Clk, bus, 0x00, [0xAA], thread_id=0x00)
+    yield a.join()
+    a = yield read(dut.Clk, bus, 0x00, 3, thread_id=0x00)
+    yield a.join()
+
+    a = yield write(dut.Clk, bus, 0x08, [0xBB], thread_id=0x01)
+    yield a.join()
+    a = yield read(dut.Clk, bus, 0x00, 3, thread_id=0x00)
+    yield a.join()
```

Comment:
This is a fairly weird API, having an coroutine that passes back another coroutine upon completion. Not too big a deal since this is only a test, but perhaps worth adding a comment. I'll have a think about whether there's a better abstraction for this.

Reply:
Yea... the multithreaded tests are a better example of why.  I'm certainly open to suggestions.